### PR TITLE
Add support for running container from OCI archive

### PR DESCRIPTION
**Pull request by austinvazquez**
### Issue
Proposal for https://github.com/containerd/nerdctl/issues/3536

### Description
This change adds the ability to create and run containers from OCI archive without performing separate load step.

### Testing
As this behavior is not provided by Docker, this change adds a nerdctl-only test for this behavior.

### CODE REVIEW

**Comment by austinvazquez** [comment]
If maintainers would prefer, I can break https://github.com/containerd/nerdctl/pull/3537/commits/79e4e24fbaaf94a13461bb48d9710a9f0c55505d into its own PR. The change was trivial at first before I added docs changes and integration test. LMK your thoughts.

File: cmd/nerdctl/container/container_create.go [hunk 1]
```diff
@@ -409,12 +409,17 @@ func processContainerCreateOptions(cmd *cobra.Command) (types.ContainerCreateOpt
 	if err != nil {
 		return opt, err
 	}
+	quiet, err := cmd.Flags().GetBool("quiet")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/container/container_create.go:412]: Should be a separate PR

Comment (AkihiroSuda) [reply cmd/nerdctl/container/container_create.go:412]: Does Docker or Podman have this flag?

Otherwise —pull-quiet might be less confusing? idk

File: cmd/nerdctl/container/container_run.go [hunk 2]
```diff
@@ -361,6 +363,33 @@ func runAction(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
+	if imageRef := args[0]; strings.HasPrefix(imageRef, "oci-archive://") {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/container/container_run.go:366]: This has to be supported in `nerdctl quiet` too (See ipfs:// code)

File: cmd/nerdctl/container/container_run.go [hunk 3]
```diff
@@ -361,6 +363,33 @@ func runAction(cmd *cobra.Command, args []string) error {
...
+			AllPlatforms: false,
+			Quiet:        createOpt.ImagePullOpt.Quiet,
+		})
+		if err != nil {
+			return err
+		}
+
+		// Running multi-image OCI archives is not yet supported.
+		if len(images) != 1 {
+			return errors.New("multiple images in OCI archive found")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/container/container_run.go:387]: What about just defaulting to the first one? (With a warning)

File: cmd/nerdctl/container/container_run_test.go [hunk 4]
```diff
@@ -631,3 +631,29 @@ func TestRunAttachFlag(t *testing.T) {
 		})
 	}
 }
+
+func TestRunFromOCIArchive(t *testing.T) {
+	testutil.RequiresBuild(t)
+	testutil.RegisterBuildCacheCleanup(t)
+	testutil.DockerIncompatible(t)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/container/container_run_test.go:638]: Needs a comment

File: docs/command-reference.md
Comment (AkihiroSuda) [inline docs/command-reference.md:1]: oci-archive://  needs socks too

File: cmd/nerdctl/container/container_create.go (hunk 1 above)
Comment (austinvazquez) [reply cmd/nerdctl/container/container_create.go:412]: 👍🏼 Opened https://github.com/containerd/nerdctl/pull/3551
Docker has --quiet across load, run, and create.

**Comment by austinvazquez** [comment]
Going to let this CI run finish before rebasing for https://github.com/containerd/nerdctl/pull/3551

File: docs/command-reference.md
Comment (austinvazquez) [reply docs/command-reference.md:1]: Addressed in https://github.com/containerd/nerdctl/pull/3537/commits/d669f3d0af9f235ce61d744c3e21b3495a9b5e69. Added documentation for both create and run.

**Review by chews93319 — APPROVED** [review]

File: pkg/cmd/container/create.go [hunk 5]
```diff
@@ -123,6 +124,37 @@ func Create(ctx context.Context, client *containerd.Client, args []string, netMa
 	}
 	opts = append(opts, platformOpts...)
 
+	if imageRef := args[0]; strings.HasPrefix(imageRef, "oci-archive://") {
```
Comment (AkihiroSuda) [inline pkg/cmd/container/create.go:127]: Protocol prefixes should be parsed in the same place as ipfs://

https://github.com/containerd/nerdctl/blob/f240e58d2e5922642128fee245d933a6550ddb46/pkg/referenceutil/referenceutil.go#L91

File: pkg/referenceutil/referenceutil_test.go [hunk 6]
```diff
@@ -290,6 +290,14 @@ func TestReferenceUtil(t *testing.T) {
 			Tag:          "latest",
 			ExplicitTag:  "",
 		},
+		"oci-archive:///tmp/build/saved-image.tar": {
+			Protocol:     "oci-archive",
+			Path:         "/tmp/build/saved-image.tar",
+			Error:        "",
+			String:       "/tmp/build/saved-image.tar",
+			Suggested:    "oci-archive-/tmp/-abcde",
```
Comment (apostasie) [inline pkg/referenceutil/referenceutil_test.go:298]: You likely need to do something for the suggested name of oci-archive://.

This suggested name is not valid for a container name (it is used when you do not specify a name for your container).

Maybe something in the line of:
```
func (ir *ImageReference) SuggestContainerName(suffix string) string {
	name := "untitled"
	if ir.Protocol != "" && ir.Domain == "" {
		if ir.Protocol == OCIArchive {
			name = string(ir.Protocol) + "-" + path.Base(ir.Path)[:shortIDLength]
		}else{
			name = string(ir.Protocol) + "-" + ir.String()[:shortIDLength]
		}
	} else if ir.Path != "" {
		name = path.Base(ir.Path)
	}
	return name + "-" + suffix[:5]
}

```

... although that is likely not sufficient, as filesystem might allow characters that are not safe to use as image identifiers.

Comment (apostasie) [reply pkg/referenceutil/referenceutil_test.go:298]: Maybe we keep it simple and only suggest ‘oci-archive-container-id’ so that we do not have to deal with local path sanitization.

File: pkg/cmd/container/create.go [hunk 7]
```diff
@@ -123,6 +124,37 @@ func Create(ctx context.Context, client *containerd.Client, args []string, netMa
...
+			return nil, nil, err
+		} else if len(images) == 0 {
+			// This is a regression and should not occur.
+			return nil, nil, errors.New("OCI archive did not contain any images")
+		}
+
+		image := images[0].Name
+		// Multi-image archive provided, default to first image found.
+		if len(images) != 1 {
+			log.L.Warnf("multi-image OCI archive provided, defaulting to image %s...", image)
```
Comment (ktock) [inline pkg/cmd/container/create.go:152]:
```suggestion
			log.L.Warnf("multiple images are found for the platform, defaulting to image %s...", image)
```

File: pkg/imgutil/load/load.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/containerd/nerdctl/v2/pkg/imgutil"
+	"github.com/containerd/nerdctl/v2/pkg/platformutil"
+)
+
+// FromArchive will load and unpack the images from the provided tar archive specified in image load options.
+func FromArchive(ctx context.Context, client *containerd.Client, options types.ImageLoadOptions) ([]images.Image, error) {
+	if options.Input != "" {
+		f, err := os.Open(options.Input)
+		if err != nil {
+			return []images.Image{}, err
```
Comment (ktock) [inline pkg/imgutil/load/load.go:43]: why not `return nil, err`?

File: pkg/referenceutil/referenceutil_test.go (hunk 6 above)
Comment (austinvazquez) [reply pkg/referenceutil/referenceutil_test.go:298]: Thanks @apostasie. Let me take another look here. We load the image from the archive before and reset the image ref to the name in the image manifest before calling parse again on that reference. I shouldn't have just made this test work though. Maybe it would be better to have another type for oci-archive image references which doesn't support suggested container name as it doesn't make sense.

File: pkg/imgutil/load/load.go (hunk 8 above)
Comment (austinvazquez) [reply pkg/imgutil/load/load.go:43]: I feel like I picked this up early in my Go learning (maybe a project I worked did this) but yeah looks like nil is recommended to avoid memory alloc.

File: pkg/referenceutil/referenceutil_test.go (hunk 6 above)
Comment (apostasie) [reply pkg/referenceutil/referenceutil_test.go:298]: I’ll have a look around as well on how this is used. 
thanks @austinvazquez !

File: pkg/imgutil/load/load.go (hunk 8 above)
Comment (austinvazquez) [reply pkg/imgutil/load/load.go]: Addressed in https://github.com/containerd/nerdctl/pull/3537/commits/d8ab512e735c20ab922198c584b8cf69ba192793

File: pkg/cmd/container/create.go (hunk 7 above)
Comment (austinvazquez) [reply pkg/cmd/container/create.go]: Applied in https://github.com/containerd/nerdctl/pull/3537/commits/932bfaf0a599fbda3f19377cc827092c371318bd

File: pkg/referenceutil/referenceutil_test.go (hunk 6 above)
Comment (austinvazquez) [reply pkg/referenceutil/referenceutil_test.go]: Should be addressed with https://github.com/containerd/nerdctl/pull/3537/commits/403499804af29636fe6168a430ee2a0ca687fc79. @apostasie, I took another look and changed how I was looking at the problem. When parsing an raw image reference which is a path for an OCI archive, we should fail and let the caller (`pkg/create`) know the reference is a path to an image archive and needs to be loaded before (re)parsing. This should resolve the quirkiness of having a image reference object which was not really a valid image reference.

File: cmd/nerdctl/container/container_create_linux_test.go [hunk 9]
```diff
@@ -306,3 +307,34 @@ func TestIssue2993(t *testing.T) {
...
+
+	// Docker does not support creating containers from OCI archive.
+	testutil.DockerIncompatible(t)
+
+	base := testutil.NewBase(t)
+	imageName := testutil.Identifier(t)
+	containerName := testutil.Identifier(t)
+
+	teardown := func() {
+		base.Cmd("rm", containerName).Run()
```
Comment (apostasie) [inline cmd/nerdctl/container/container_create_linux_test.go:323]: `-f` for both rm and rmi would be better in teardown.

File: cmd/nerdctl/container/container_run_test.go [hunk 10]
```diff
@@ -658,3 +658,31 @@ func TestRunQuiet(t *testing.T) {
...
+	testutil.RegisterBuildCacheCleanup(t)
+
+	// Docker does not support running container images from OCI archive.
+	testutil.DockerIncompatible(t)
+
+	base := testutil.NewBase(t)
+	imageName := testutil.Identifier(t)
+
+	teardown := func() {
+		base.Cmd("rmi", imageName).Run()
```
Comment (apostasie) [inline cmd/nerdctl/container/container_run_test.go:673]: -f

File: pkg/referenceutil/referenceutil_test.go [hunk 11]
```diff
@@ -273,6 +273,10 @@ func TestReferenceUtil(t *testing.T) {
 			Tag:          "latest",
 			ExplicitTag:  "",
 		},
+		"oci-archive:///tmp/build/saved-image.tar": {
+			Protocol: "oci-archive",
```
Comment (apostasie) [inline pkg/referenceutil/referenceutil_test.go:277]: Protocol is not necessary here, since the test stops as it errors (and since your latest version does not set the protocol)

File: pkg/referenceutil/referenceutil_test.go (hunk 6 above)
Comment (apostasie) [reply pkg/referenceutil/referenceutil_test.go]: Yeah, reading through - it is a bit weird indeed, because, yep, it is not really an image ref, it is a filesystem location... - I think it is ok as you put it now.

**Review by apostasie — APPROVED** [review]

**Comment by apostasie** [comment]
@austinvazquez thanks! - I left a few nits comments.

**Review by AkihiroSuda — APPROVED** [review]
Thanks
